src/core/expr/fexpr_cut.cc

```diff
@@ -110,17 +134,120 @@ class FExpr_Cut : public FExpr_Func {
         }
       }
 
-      // Cut workframe in-place
+      // Bin columns in-place
       for (size_t i = 0; i < ncols; ++i) {
-        Column coli = wf.retrieve_column(i);
-        coli = Column(Cut_ColumnImpl::make(
-                 std::move(coli), i, nbins[i], right_closed_
-               ));
-        wf.replace_column(i, std::move(coli));
+        Column col = wf.retrieve_column(i);
+
+        if (col.ltype() == LType::MU) {
+          col = Column(new ConstNa_ColumnImpl(col.nrows(), dt::SType::INT32));
+        } else {
+          if (!ltype_is_numeric(col.ltype())) {
+            throw TypeError() << "cut() can only be applied to numeric or void "
+              << "columns, instead column `" << i << "` has an stype: `"
+              << col.stype() << "`";
+          }
+
+          col = Column(CutNbins_ColumnImpl::make(
+                  std::move(col), nbins[i], right_closed_
+                ));
+        }
+        wf.replace_column(i, std::move(col));
       }
+    }
 
-      return wf;
+
+    // Bin data based on the provided interval edges.
+    void cut_bins(Workframe& wf) const {
+      const size_t ncols = wf.ncols();
+
+      if (!py_bins_.is_list_or_tuple()) {
+        throw TypeError() << "`bins` parameter must be a list or a tuple, "
+          << "instead got `" << py_bins_.typeobj() << "`";
+      }
+
+      py::oiter py_bins = py_bins_.to_oiter();
+
+      if (py_bins.size() != ncols) {
+        throw ValueError() << "Number of elements in `bins` must be equal to "
+          << "the number of columns in the frame/expression, i.e. `"
+          << ncols << "`, instead got: `" << py_bins.size() << "`";
+      }
+
+      size_t i = 0;
+      for (auto py_bin : py_bins) {
+        DataTable* dt = py_bin.to_datatable();
+        if (dt->ncols() != 1) {
+          throw ValueError() << "To bin a column `cut()` needs exactly one column "
+            << "with the bin edges, instead for the frame `" << i << "` got: "
+            << dt->ncols() << "`";
+        }
+
+        if (dt->nrows() < 2) {
+          throw ValueError() << "To bin data at least two edges are required, "
+            << "instead for the frame `" << i << "` got: `" << dt->nrows() << "`";
+        }
+
+        // Retrieve bins
+        Column bins = dt->get_column(0);
+        if (!ltype_is_numeric(bins.ltype())) {
+          throw TypeError() << "Bin edges must be provided as the numeric columns only, "
+            << "instead for the frame `" << i << "` the column stype is `"
+            << bins.stype() << "`";
+        }
+        bins.cast_inplace(SType::FLOAT64);
+        validate_bins(bins, i);
```

Comment:
Validation  should happen as soon as possible: in this case the validity of bins could have been checked as soon as this FExpr is created.
If the fexpr is reused you're currently re-doing the validation every time.

Reply:
Yes, not sure why but I have placed all the validation to `evaluate_n()` from the very beginning. It is not just about `validate_bins()`. Will probably move it to `pyfn_cut()` and also `pyfn_qcut()`...

Reply:
I've been thinking about why I first put all the validation into the `evaluate_n()` and not to the corresponding `pyfn_` functions and found that probably I've reused the pattern we already had. `ifelse`, `round`, etc. they all perform validation at the `evaluate_n` stage. Any reason for that?